pkg/gatewayserver/io/basicstationlns/messages/upstream.go

```diff
@@ -0,0 +1,288 @@
+		Alias: Alias(req),
+	})
+}
+
+// UplinkDataFrame is the LoRaWAN Uplink message from the BasicStation.
+type UplinkDataFrame struct {
+	MHdr          uint    `json:"MHdr"`
+	DevAddr       int32   `json:"DevAddr"`
+	FCtrl         uint    `json:"FCtrl"`
+	FCnt          uint    `json:"Fcnt"`
+	FOpts         string  `json:"FOpts"`
+	FPort         int     `json:"FPort"`
+	FRMPayload    string  `json:"FRMPayload"`
+	MIC           int32   `json:"MIC"`
+	RefTime       float64 `json:"RefTime"`
+	RadioMetaData RadioMetaData
+}
+
+// MarshalJSON implements json.Marshaler.
+func (updf UplinkDataFrame) MarshalJSON() ([]byte, error) {
+	type Alias UplinkDataFrame
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
+	}{
+		Type:  TypeUpstreamUplinkDataFrame,
+		Alias: Alias(updf),
+	})
+}
+
+// TxConfirmation is the LoRaWAN Join Request message from the BasicStation.
+type TxConfirmation struct {
+	Diid    int64            `json:"diid"`
+	DevEUI  basicstation.EUI `json:"DevEui"`
+	RCtx    int64            `json:"rctx"`
+	XTime   int64            `json:"xtime"`
+	TxTime  float64          `json:"txtime"`
+	GpsTime int64            `json:"gpstime"`
+}
+
+// MarshalJSON implements json.Marshaler.
+func (conf TxConfirmation) MarshalJSON() ([]byte, error) {
+	type Alias TxConfirmation
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
+	}{
+		Type:  TypeUpstreamTxConfirmation,
+		Alias: Alias(conf),
+	})
+}
+
+// ToUplinkMessage extracts fields from the basic station Join Request "jreq" message and converts them into an UplinkMessage for the network server.
+func (req *JoinRequest) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (*ttnpb.UplinkMessage, error) {
+	var up ttnpb.UplinkMessage
+	up.ReceivedAt = time.Now()
+
+	var parsedMHDR ttnpb.MHDR
+	if err := lorawan.UnmarshalMHDR([]byte{byte(req.MHdr)}, &parsedMHDR); err != nil {
+		return nil, errJoinRequestMessage.WithCause(err)
+	}
+
+	micBytes, err := getInt32AsByteSlice(req.MIC)
+	if err != nil {
+		return nil, errJoinRequestMessage.WithCause(err)
+	}
+	up.Payload = &ttnpb.Message{
+		MHDR: parsedMHDR,
+		MIC:  micBytes,
+		Payload: &ttnpb.Message_JoinRequestPayload{JoinRequestPayload: &ttnpb.JoinRequestPayload{
+			JoinEUI:  req.JoinEUI.EUI64,
+			DevEUI:   req.DevEUI.EUI64,
+			DevNonce: [2]byte{byte(req.DevNonce), byte(req.DevNonce >> 8)},
+		}},
+	}
+
+	up.RawPayload, err = lorawan.MarshalMessage(*up.Payload)
+	if err != nil {
+		return nil, errJoinRequestMessage.WithCause(err)
+	}
+
+	timestamp := uint32(req.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
+
+	ulToken := ttnpb.UplinkToken{
+		GatewayAntennaIdentifiers: ttnpb.GatewayAntennaIdentifiers{
+			GatewayIdentifiers: ids,
+			AntennaIndex:       uint32(req.RadioMetaData.UpInfo.RCtx),
+		},
+		Timestamp: timestamp,
+	}
+	ulTokenBytes, err := ulToken.Marshal()
+	if err != nil {
+		return nil, errJoinRequestMessage.WithCause(err)
+	}
+
+	rxTime := time.Unix(req.RadioMetaData.UpInfo.RxTime, 0)
```

Comment:
These are not seconds right?

Reply:
`RxTime` is not defined in the official spec but is sent by the gateway in float64 UTC. Updated to reflect that.